File: NovinAI-SPM/Sources/NovinAI/ProtectedPython/feature_extractor.py
"""Feature engineering pipeline for NovinAI security events."""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from datetime import datetime, timezone
from functools import lru_cache
from typing import Any, Dict, Iterable, List, Mapping, MutableMapping, Optional

import mmh3
import numpy as np

logger = logging.getLogger(__name__)


@dataclass(slots=True)
class FeatureConfig:
    max_features: int = 16_384
    time_window_hours: int = 24
    location_precision: int = 3
    event_types: tuple[str, ...] = (
        "motion",
        "sound",
        "door",
        "window",
        "face",
        "smoke",
        "fire",
        "glassbreak",
        "pet",
        "vehicle",
    )
    feature_scaling: bool = True
    dtype: np.dtype = field(default=np.float32, repr=False)


class FeatureExtraction:
    """Transforms heterogeneous sensor payloads into dense feature vectors."""

    def __init__(self, config: Any) -> None:
        self.config = config
        self.feature_config = FeatureConfig()
        self.scaling_params: Dict[str, np.ndarray] = {}

    def extract(self, request_data: Dict[str, Any], crime_context: Dict[str, Any]) -> np.ndarray:
        try:
            features: Dict[str, float] = {}
            features.update(self._extract_temporal_features(request_data))
            features.update(self._extract_spatial_features(request_data, crime_context))
            features.update(self._extract_event_features(request_data))
            features.update(self._extract_behavioral_features(request_data))
            features.update(self._extract_environmental_features(request_data, crime_context))

            vector = self._vectorize_features(features)
            if self.feature_config.feature_scaling:
                vector = self._scale_features(vector)
            return vector
        except Exception as exc:
            logger.exception("Feature extraction failure: %s", exc)
            return np.zeros(self.feature_config.max_features, dtype=self.feature_config.dtype)

    # ------------------------------------------------------------------
    # Individual feature families
    # ------------------------------------------------------------------
    def _extract_temporal_features(self, request_data: Mapping[str, Any]) -> Dict[str, float]:
        timestamp = request_data.get("timestamp", datetime.now(timezone.utc))
        if isinstance(timestamp, str):
            try:
                timestamp = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            except ValueError:
                timestamp = datetime.now(timezone.utc)

        hour = timestamp.hour
        weekday = timestamp.weekday()
        month = timestamp.month

        features = {
            "hour_sin": float(np.sin(2 * np.pi * hour / 24)),
            "hour_cos": float(np.cos(2 * np.pi * hour / 24)),
            "weekday_sin": float(np.sin(2 * np.pi * weekday / 7)),
            "weekday_cos": float(np.cos(2 * np.pi * weekday / 7)),
            "month_sin": float(np.sin(2 * np.pi * month / 12)),
            "month_cos": float(np.cos(2 * np.pi * month / 12)),
            "is_weekend": float(weekday >= 5),
        }

        last_event = request_data.get("last_event_time")
        if last_event:
            if isinstance(last_event, str):
                try:
                    last_event = datetime.fromisoformat(last_event.replace("Z", "+00:00"))
                except ValueError:
                    last_event = None
            if isinstance(last_event, datetime):
                delta = (timestamp - last_event).total_seconds() / 3600
                features["hours_since_last_event"] = float(np.clip(delta, 0, 24) / 24)
        else:
            features["hours_since_last_event"] = 1.0
        return features

    def _extract_spatial_features(self, request_data: Mapping[str, Any], crime_context: Mapping[str, Any]) -> Dict[str, float]:
        location = request_data.get("location", {})
        lat = float(location.get("latitude", 0.0))
        lng = float(location.get("longitude", 0.0))

        features = {
            "latitude_norm": (lat + 90.0) / 180.0,
            "longitude_norm": (lng + 180.0) / 360.0,
            "crime_rate_24h": float(crime_context.get("crime_rate_24h", 0.0)),
            "crime_rate_7d": float(crime_context.get("crime_rate_7d", 0.0)),
            "crime_rate_30d": float(crime_context.get("crime_rate_30d", 0.0)),
            "nearby_incidents": min(float(crime_context.get("nearby_incidents", 0.0)), 20.0) / 20.0,
            "crime_severity": float(crime_context.get("avg_severity", 0.0)),
        }
        return features

    def _extract_event_features(self, request_data: Mapping[str, Any]) -> Dict[str, float]:
        event_type = request_data.get("event_type", "unknown")
        event_data = request_data.get("event_data", {})

        features = {f"event_{etype}": float(event_type == etype) for etype in self.feature_config.event_types}
        features["event_confidence"] = float(event_data.get("confidence", 0.5))
        features["event_duration"] = float(min(event_data.get("duration", 0), 600) / 600)
        features["event_intensity"] = float(event_data.get("intensity", 0.5))
        features["sensor_count"] = float(min(len(event_data.get("sensors_triggered", [])), 6) / 6)
        return features

    def _extract_behavioral_features(self, request_data: Mapping[str, Any]) -> Dict[str, float]:
        history = request_data.get("activity_history", [])[-20:]
        risk_profile = request_data.get("user_risk_profile", {})

        features = {
            "recent_activity_freq": float(min(len(history), 20) / 20.0),
            "user_risk_score": float(risk_profile.get("risk_score", 0.5)),
            "user_trust_level": float(risk_profile.get("trust_level", 0.5)),
        }

        if history:
            hours = [entry.get("hour", 12) for entry in history]
            features["activity_consistency"] = float(1.0 - (np.std(hours) / 12.0))
        else:
            features["activity_consistency"] = 0.5
        return features

    def _extract_environmental_features(self, request_data: Mapping[str, Any], crime_context: Mapping[str, Any]) -> Dict[str, float]:
        weather = request_data.get("weather", {})
        timestamp = request_data.get("timestamp", datetime.now(timezone.utc))
        if isinstance(timestamp, str):
            try:
                timestamp = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            except ValueError:
                timestamp = datetime.now(timezone.utc)

        hour = timestamp.hour
        month = timestamp.month
        season_index = (month % 12) // 3

        features = {
            "temperature": float((weather.get("temperature", 21) + 40) / 80),
            "humidity": float(weather.get("humidity", 50) / 100),
            "precipitation": float(min(weather.get("precipitation", 0), 50) / 50),
            "wind_speed": float(min(weather.get("wind_speed", 0), 40) / 40),
            "is_daylight": float(6 <= hour <= 18),
        }

        seasons = ["season_winter", "season_spring", "season_summer", "season_fall"]
        for idx, key in enumerate(seasons):
            features[key] = float(idx == season_index)
        return features

    # ------------------------------------------------------------------
    # Vectorisation & scaling
    # ------------------------------------------------------------------
    def _vectorize_features(self, features: Mapping[str, float]) -> np.ndarray:
        vector = np.zeros(self.feature_config.max_features, dtype=self.feature_config.dtype)
        for name, value in features.items():
            slot = self._feature_slot(name)
            vector[slot] = float(value)
        return vector

    @lru_cache(maxsize=10_000)
    def _feature_slot(self, feature_name: str) -> int:
        return int(mmh3.hash(feature_name) % self.feature_config.max_features)

    def _scale_features(self, vector: np.ndarray) -> np.ndarray:
        if not self.scaling_params:
            self.scaling_params = {
                "mean": np.zeros_like(vector),
                "std": np.ones_like(vector),
            }
        mean = self.scaling_params["mean"]
        std = self.scaling_params["std"]
        scaled = (vector - mean) / (std + 1e-6)
        return np.clip(scaled, -5.0, 5.0)

    # ------------------------------------------------------------------
    # Diagnostics
    # ------------------------------------------------------------------
    def get_feature_names(self) -> List[str]:
        return list(self.feature_config.event_types)


__all__ = ["FeatureExtraction", "FeatureConfig"]

File: Sources/NovinIntelligence/CoreML/FeatureExtractor.swift
import Foundation

@available(iOS 15.0, macOS 12.0, *)
struct FeatureExtractorSwift {
    struct Config {
        static let maxFeatures = 16_384
    }

    // MARK: - Constants & caches
    private static let hoursInDay = 24
    private static let daysInWeek = 7
    private static let monthsInYear = 12
    private static let hourSin: [Double] = (0..<hoursInDay).map { sin(2 * .pi * Double($0) / Double(hoursInDay)) }
    private static let hourCos: [Double] = (0..<hoursInDay).map { cos(2 * .pi * Double($0) / Double(hoursInDay)) }
    private static let weekdaySin: [Double] = (0..<daysInWeek).map { sin(2 * .pi * Double($0) / Double(daysInWeek)) }
    private static let weekdayCos: [Double] = (0..<daysInWeek).map { cos(2 * .pi * Double($0) / Double(daysInWeek)) }
    private static let monthSin: [Double] = (0..<monthsInYear).map { sin(2 * .pi * Double($0 + 1) / Double(monthsInYear)) }
    private static let monthCos: [Double] = (0..<monthsInYear).map { cos(2 * .pi * Double($0 + 1) / Double(monthsInYear)) }

    // Public API
    func extract(from request: [String: Any]) -> [Float] {
        var feats: [String: Double] = [:]

        // Temporal
        addTemporalFeatures(from: request, into: &feats)
        // Spatial
        addSpatialFeatures(from: request, into: &feats)
        // Event features (supports either top-level fields or first item under "events")
        let event = extractPrimaryEvent(from: request)
        addEventFeatures(from: event, into: &feats)
        // Behavioral
        addBehavioralFeatures(from: request, into: &feats)
        // Environmental
        addEnvironmentalFeatures(from: request, into: &feats)

        // Vectorize via MurmurHash3-32 slots
        var vector = [Float](repeating: 0, count: Config.maxFeatures)
        for (name, value) in feats {
            let slot = featureSlot(name)
            vector[slot] = Float(value)
        }
        return vector
    }

    /// New: Extract named features for math+rules pipeline
    func extractNamedFeatures(from request: [String: Any]) -> [String: Double] {
        var feats: [String: Double] = [:]
        // Temporal
        addTemporalFeatures(from: request, into: &feats)
        // Spatial
        addSpatialFeatures(from: request, into: &feats)
        // Event
        let event = extractPrimaryEvent(from: request)
        addEventFeatures(from: event, into: &feats)
        // Behavioral
        addBehavioralFeatures(from: request, into: &feats)
        // Environmental
        addEnvironmentalFeatures(from: request, into: &feats)
        return feats
    }

    // MARK: - Families
    private func addTemporalFeatures(from req: [String: Any], into out: inout [String: Double]) {
        let timestamp: TimeInterval = {
            if let t = req["timestamp"] as? TimeInterval { return t }
            if let s = req["timestamp"] as? String {
                let iso = ISO8601DateFormatter()
                if let d = iso.date(from: s) { return d.timeIntervalSince1970 }
            }
            return currentUnixTime()
        }()
        let date = Date(timeIntervalSince1970: timestamp)
        let cal = Calendar(identifier: .gregorian)
        let hour = cal.component(.hour, from: date)
        let weekday = cal.component(.weekday, from: date) - 1 // 0..6
        let month = cal.component(.month, from: date)

        out["hour_sin"] = Self.hourSin[hour]
        out["hour_cos"] = Self.hourCos[hour]
        let wd = max(0, min(Self.daysInWeek - 1, weekday))
        out["weekday_sin"] = Self.weekdaySin[wd]
        out["weekday_cos"] = Self.weekdayCos[wd]
        let mi = max(1, min(Self.monthsInYear, month)) - 1
        out["month_sin"] = Self.monthSin[mi]
        out["month_cos"] = Self.monthCos[mi]
        out["is_weekend"] = (weekday >= 5) ? 1.0 : 0.0

        if let lastT = req["last_event_time"] {
            let last: TimeInterval? = (lastT as? TimeInterval) ?? (lastT as? String).flatMap { ISO8601DateFormatter().date(from: $0)?.timeIntervalSince1970 }
            let deltaHrs = max(0, (timestamp - (last ?? timestamp)) / 3600.0)
            out["hours_since_last_event"] = min(deltaHrs, 24) / 24.0
        } else {
            out["hours_since_last_event"] = 1.0
        }
    }

    private func addSpatialFeatures(from req: [String: Any], into out: inout [String: Double]) {
        let loc = (req["location"] as? [String: Any]) ?? [:]
        let latRaw = (loc["latitude"] as? Double) ?? 0
        let lngRaw = (loc["longitude"] as? Double) ?? 0
        let lat = max(-90.0, min(90.0, latRaw))
        let lng = max(-180.0, min(180.0, lngRaw))

        let crime = (req["crime_context"] as? [String: Any]) ?? [:]

        out["latitude_norm"] = (lat + 90.0) / 180.0
        out["longitude_norm"] = (lng + 180.0) / 360.0
        out["crime_rate_24h"] = (crime["crime_rate_24h"] as? Double) ?? 0
        out["crime_rate_7d"] = (crime["crime_rate_7d"] as? Double) ?? 0
        out["crime_rate_30d"] = (crime["crime_rate_30d"] as? Double) ?? 0
        out["nearby_incidents"] = min(((crime["nearby_incidents"] as? Double) ?? 0), 20.0) / 20.0
        out["crime_severity"] = (crime["avg_severity"] as? Double) ?? 0
    }

    private func addEventFeatures(from event: [String: Any], into out: inout [String: Double]) {
        let rawType = ((event["type"] as? String) ?? (event["event_type"] as? String) ?? "unknown").lowercased()
        let type: String = {
            // Normalize common vendor-specific types
            if rawType.contains("motion") { return "motion" }
            if rawType.contains("glass") { return "glassbreak" }
            return rawType
        }()
        let data = (event["event_data"] as? [String: Any]) ?? (event["metadata"] as? [String: Any]) ?? [:]
        let eventTypes = ["motion","sound","door","window","face","smoke","fire","glassbreak","pet","vehicle"]
        for et in eventTypes {
            out["event_\(et)"] = (type == et) ? 1.0 : 0.0
        }
        // Derive specific subtypes
        if let soundType = (data["sound_type"] as? String)?.lowercased(), soundType.contains("glass") {
            out["event_glassbreak"] = 1.0
        }

        // Clamp numeric fields into [0,1] or normalized ranges
        let confidence = (data["confidence"] as? Double) ?? (event["confidence"] as? Double) ?? 0.5
        out["event_confidence"] = max(0.0, min(1.0, confidence))
        let duration = max(0.0, (data["duration"] as? Double) ?? 0)
        out["event_duration"] = min(duration, 600) / 600.0
        let intensity = (data["intensity"] as? Double) ?? 0.5
        out["event_intensity"] = max(0.0, min(1.0, intensity))
        if let sensors = data["sensors_triggered"] as? [Any] {
            out["sensor_count"] = min(Double(sensors.count), 6) / 6.0
        } else {
            out["sensor_count"] = 0.0
        }
    }

    private func addBehavioralFeatures(from req: [String: Any], into out: inout [String: Double]) {
        let hist = (req["activity_history"] as? [[String: Any]]) ?? []
        let tail = Array(hist.suffix(20))
        let risk = (req["user_risk_profile"] as? [String: Any]) ?? [:]

        out["recent_activity_freq"] = min(Double(tail.count), 20.0) / 20.0
        out["user_risk_score"] = (risk["risk_score"] as? Double) ?? 0.5
        out["user_trust_level"] = (risk["trust_level"] as? Double) ?? 0.5

        if !tail.isEmpty {
            let hours = tail.compactMap { $0["hour"] as? Double }
            if hours.count > 1 {
                let mean = hours.reduce(0, +) / Double(hours.count)
                // sample variance when n>1
                let variance = hours.reduce(0) { $0 + pow($1 - mean, 2) } / Double(hours.count - 1)
                let std = sqrt(max(0, variance))
                let score = 1.0 - (std / 12.0)
                out["activity_consistency"] = max(0.0, min(1.0, score))
            } else {
                out["activity_consistency"] = 0.5
            }
        } else {
            out["activity_consistency"] = 0.5
        }
    }

    private func addEnvironmentalFeatures(from req: [String: Any], into out: inout [String: Double]) {
        let weather = (req["weather"] as? [String: Any]) ?? [:]
        let timestamp = (req["timestamp"] as? TimeInterval) ?? currentUnixTime()
        let date = Date(timeIntervalSince1970: timestamp)
        let cal = Calendar(identifier: .gregorian)
        let hour = cal.component(.hour, from: date)
        let month = cal.component(.month, from: date)
        let seasonIndex = ((month % 12) / 3)

        out["temperature"] = (((weather["temperature"] as? Double) ?? 21) + 40.0) / 80.0
        out["humidity"] = ((weather["humidity"] as? Double) ?? 50) / 100.0
        out["precipitation"] = min(((weather["precipitation"] as? Double) ?? 0), 50.0) / 50.0
        out["wind_speed"] = min(((weather["wind_speed"] as? Double) ?? 0), 40.0) / 40.0
        out["is_daylight"] = ((6 <= hour && hour <= 18) ? 1.0 : 0.0)

        let seasons = ["season_winter", "season_spring", "season_summer", "season_fall"]
        for (idx, key) in seasons.enumerated() {
            out[key] = (idx == seasonIndex) ? 1.0 : 0.0
        }
    }

    // MARK: - Event selection
    private func extractPrimaryEvent(from req: [String: Any]) -> [String: Any] {
        if let events = req["events"] as? [[String: Any]], let first = events.first {
            return first
        }
        return req
    }

    // MARK: - Hashing & slots (stable across runs; Murmur3 x86_32, seed=0)
    private func featureSlot(_ name: String) -> Int {
        let h = murmurHash3_32(name)
        return Int(h % UInt32(Config.maxFeatures))
    }

    private func currentUnixTime() -> TimeInterval { Date().timeIntervalSince1970 }
}
 
// MARK: - Utilities
private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self { min(max(self, range.lowerBound), range.upperBound) }
}

// MARK: - MurmurHash3 (x86_32) with fixed seed for determinism
@available(iOS 15.0, macOS 12.0, *)
private func murmurHash3_32(_ s: String, seed: UInt32 = 0) -> UInt32 {
    let c1: UInt32 = 0xcc9e2d51
    let c2: UInt32 = 0x1b873593
    let r1: UInt32 = 15
    let r2: UInt32 = 13
    let m: UInt32 = 5
    let n: UInt32 = 0xe6546b64

    var hash = seed
    let data = Array(s.utf8)
    let len = data.count
    let nblocks = len / 4

    // body
    for i in 0..<nblocks {
        let i4 = i * 4
        var k: UInt32 = 0
        k |= UInt32(data[i4 + 0])
        k |= UInt32(data[i4 + 1]) << 8
        k |= UInt32(data[i4 + 2]) << 16
        k |= UInt32(data[i4 + 3]) << 24

        k &*= c1
        k = (k << r1) | (k >> (32 - r1))
        k &*= c2

        hash ^= k
        hash = ((hash << r2) | (hash >> (32 - r2)))
        hash = hash &* m &+ n
    }

    // tail
    var k1: UInt32 = 0
    let tailIndex = nblocks * 4
    let remaining = len & 3
    if remaining == 3 { k1 ^= UInt32(data[tailIndex + 2]) << 16 }
    if remaining >= 2 { k1 ^= UInt32(data[tailIndex + 1]) << 8 }
    if remaining >= 1 {
        k1 ^= UInt32(data[tailIndex + 0])
        k1 &*= c1
        k1 = (k1 << r1) | (k1 >> (32 - r1))
        k1 &*= c2
        hash ^= k1
    }

    // finalization
    hash ^= UInt32(len)

    // fmix
    hash ^= hash >> 16
    hash &*= 0x85ebca6b
    hash ^= hash >> 13
    hash &*= 0xc2b2ae35
    hash ^= hash >> 16

    return hash
}
